pkg/controller/pkg/revision/reconciler.go

```diff
@@ -0,0 +1,565 @@
+
+// WithInstallNamespace determines the namespace where install jobs will be
+// created.
+func WithInstallNamespace(n string) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.namespace = n
+	}
+}
+
+// WithNewPackageRevisionFn determines the type of package being reconciled.
+func WithNewPackageRevisionFn(f func() v1alpha1.PackageRevision) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.newPackageRevision = f
+	}
+}
+
+// WithLogger specifies how the Reconciler should log messages.
+func WithLogger(log logging.Logger) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.log = log
+	}
+}
+
+// WithRecorder specifies how the Reconciler should record Kubernetes events.
+func WithRecorder(er event.Recorder) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.record = er
+	}
+}
+
+// WithParser specifies how the Reconciler should parse a package.
+func WithParser(p parser.Parser) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.parser = p
+	}
+}
+
+// WithParserBackend specifies how the Reconciler should parse a package.
+func WithParserBackend(p parser.Backend) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.backend = p
+	}
+}
+
+// BuildMetaScheme builds the default scheme used for identifying metadata in a
+// Crossplane package.
+func BuildMetaScheme() (*runtime.Scheme, error) {
+	metaScheme := runtime.NewScheme()
+	if err := pkgmeta.SchemeBuilder.AddToScheme(metaScheme); err != nil {
+		return nil, err
+	}
+	return metaScheme, nil
+}
+
+// BuildObjectScheme builds the default scheme used for identifying objects in a
+// Crossplane package.
+func BuildObjectScheme() (*runtime.Scheme, error) {
+	objScheme := runtime.NewScheme()
+	if err := apiextensionsv1alpha1.AddToScheme(objScheme); err != nil {
+		return nil, err
+	}
+	if err := v1beta1.AddToScheme(objScheme); err != nil {
+		return nil, err
+	}
+	return objScheme, nil
+}
+
+// Reconciler reconciles packages.
+type Reconciler struct {
+	namespace  string
+	client     resource.ClientApplicator
+	kubeClient *kubernetes.Clientset
+	parser     parser.Parser
+	backend    parser.Backend
+	log        logging.Logger
+	record     event.Recorder
+
+	newPackageRevision func() v1alpha1.PackageRevision
+}
+
+// SetupProviderRevision adds a controller that reconciles ProviderRevisions.
+func SetupProviderRevision(mgr ctrl.Manager, l logging.Logger) error {
+	name := "packages/" + strings.ToLower(v1alpha1.ProviderRevisionGroupKind)
+	nr := func() v1alpha1.PackageRevision { return &v1alpha1.ProviderRevision{} }
+
+	cfg, err := config.GetConfig()
+	if err != nil {
+		return errors.Wrap(err, "failed to initialize host config with in cluster config")
+	}
+
+	clientset, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		return errors.Wrap(err, "failed to initialize host clientset with in cluster config")
+	}
+
+	installNamespace := os.Getenv(envPodNamespace)
+	if installNamespace == "" {
+		return errors.New("cannot initialize controller with empty install namespace")
+	}
```

Comment:
Would it be possible to take this as a CLI flag rather than an environment variable? I feel that flags are somewhat more discoverable, and kingpin lets us set them from an env var so they're sort of the best of both worlds. They're slightly annoying to plumb down, but there is precedent; https://github.com/crossplane/crossplane/blob/0b4227e/cmd/crossplane/package/manage/manage.go#L95

Reply:
@negz yep! Works for me. I went with env var for testing purposes, but if we are plumbing it down then that works even better 👍